File: Java/sort/CocktailSort.java
import java.util.*;

public class CocktailSort {

    private static void swap(int arr[], int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    private static void cocktailSort(int arr[]) {
        boolean isSwapped = true;
        int begin = 0;
        int last = arr.length-1;

        while (isSwapped == true) {
            // set isSwapped to false so that if nothing gets
            // sorted in this iteration we can break right away
            isSwapped = false;

            for(int i = begin; i < last; i++) {
                // going forward
                if (arr[i] > arr[i+1]) {
                    swap(arr, i, i+1);
                    isSwapped = true;
                }
            }

            // if nothing swapped then the array is already sorted
            if (isSwapped == false)
                break;

            // else, reset isSwapped to check for next iteration
            isSwapped = false;

            // since largest item is now in last. we only need
            // to check for one place before it
            last -= 1;

            for (int i = last; i >= begin; i--) {
                // going backward
                if (arr[i] > arr[i+1]) {
                    swap(arr, i, i+1);
                    isSwapped = true;
                }
            }

            // now the smallest number is on first place
            // so move starting point one position ahead
            begin++;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        //taking input array
        System.out.println("Enter size of array:");
        int size = sc.nextInt();
        int arr[] = new int[size];
        System.out.println("Enter array elements:");
        for (int i=0; i<size; i++) {
            arr[i] = sc.nextInt();
        }
        sc.close();

        // before sorting
        System.out.println("Array before Cocktail sort:");
        for (int i=0; i<arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();

        cocktailSort(arr);

        // after sorting
        System.out.println("Array after Cocktail sort:");
        for (int i=0; i<arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
    }
}
/**
 * Sample input/output
 * Enter size of array:
 * 10
 * Enter array elements:
 * 88 26 77 49 91 62 33 85 99 11
 * Array before Cocktail sort:
 * 88 26 77 49 91 62 33 85 99 11 
 * Array after Cocktail sort:
 * 11 26 33 49 62 77 85 88 91 99
 * 
 * Enter size of array:
 * 7
 * Enter array elements:
 * 5 1 4 2 8 0 2
 * Array before Cocktail sort:
 * 5 1 4 2 8 0 2 
 * Array after Cocktail sort:
 * 0 1 2 2 4 5 8
 * 
 * Worst case Time complexity = O(n*n)
 * Best case Time complexity = O(n) when array is already sorted
 * Auxillary space = O(1)
 */

File: Python/sort/cocktail_sort.py
# Cocktail sort in python


def cocktailSort(arr):
    isSwapped = True
    First = 0
    Last = len(arr)-1
    while isSwapped is True:
        # reset isSwapped so that we can use it for this iteration
        isSwapped = False
        # traversing from first to last element
        for i in range(First, Last):
            if arr[i] > arr[i+1]:
                arr[i], arr[i+1] = arr[i+1], arr[i]
                isSwapped = True

        # if array is already sorted, break the loop
        if isSwapped is False:
            break

        isSwapped = False
        # last element is largest so we'll reduce last by one position.
        Last = Last-1
        # iteration from last to first element
        for i in range(Last-1, First-1, -1):
            if arr[i] > arr[i+1]:
                arr[i], arr[i+1] = arr[i+1], arr[i]
                isSwapped = True
        # first element is smallest so moving a first one position ahead
        First = First+1


# number of elements
n = int(input("Enter number of elements : "))
arr = list(map(int, input("\nEnter the numbers : ").strip().split()))[:n]
cocktailSort(arr)
print("Sorted array is: ")
print(arr)


'''output:

Time complexity: O(n*n)
Space: O(1)

Enter number of elements : 5
Enter the numbers : 12 4 5 34 2
Sorted array is:
[2, 4, 5, 12, 34]

Enter number of elements : 3
Enter the numbers : 20 14 34
Sorted array is:
[14, 20, 34]
'''

